src/Storages/MergeTree/MergeTreeDataMergerMutator.cpp

```diff
@@ -414,19 +351,6 @@ std::expected<MergeSelectorChoice, SelectMergeFailure> MergeTreeDataMergerMutato
         return std::move(merge_choice.value());
     }
 
-    const auto partitions_stats = calculateStatisticsForPartitions(ranges);
-
-    if (auto best = getBestPartitionToOptimizeEntire(selector.max_total_size_to_merge, context, settings, partitions_stats, log); !best.empty())
-    {
-        return selectAllPartsToMergeWithinPartition(
-            metadata_snapshot,
-            parts_collector,
-            merge_predicate,
-            /*partition_id=*/best,
-            /*final=*/true,
-            /*optimize_skip_merged_partitions=*/true);
-    }
```

Comment:
> To make this work, I had to decouple the regular selection of parts to merge from selecting an entire partition to merge. 

Why? Looks like we have the settings here, and we can add the `cleanup` flag to `MergeSelectorChoice` (or add a new `MergeType`). This would avoid duplicating code for MT/RMT/SMT

Reply:
You're right. I remember there was some reason it wouldn't work, but I got it to work now. Will update with a simpler version.